### ISSUE TITLE
Make flag sniffing optional

### ISSUE THREAD

**Issue by bolinfest** [chalk/supports-color#40]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2440%20Funded-%2300A156.svg" />](https://issuehunt.io/r/chalk/supports-color/issues/40)

Although the heuristics for inspecting `process.argv` via `hasFlag()` are pretty reasonable, they are by no means foolproof and can lead to false positives and negatives. It would be nice if it were possible to use the functionality of `supportLevel()` without any of the `hasFlag()` logic. I believe the remaining logic would be a function of:
- `process.env.CI`
- `process.env.TEAMCITY_VERSION`
- `process.env. COLORTERM`
- `process.env.TERM`
- `process.stdout`
- `process.platform`

It would be even more flexible if it were possible to pass in the `process` object so that it could be mocked and/or used from another process.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/chalk/supports-color/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by isaacs** [comment]
Another use-case: in node-tap you can pass `--color` to enable colors, and `-c` is a shorthand.  I'd like to add support for `--color=256` _and also_ for `-c256` as a shorthand, but if that means I'd have to either duplicate the work of supports-color, or conflict with it.

Also, reading the env for things like `FORCE_COLORS` is a nice default, but it can be weird.  Most utilities have their own "turn colors on" type of environment variable, which is useful since you may not want an uncolored `ls` to pipe into a colored `grep` or whatever.

Basically, my preference would be for this module to determine whether colors should be by-default *supported*, but not whether the user has explicitly requested that they be turned on.

("Well, then, Isaac, go make your own module for that" is a perfectly acceptable answer.  The space for solutions is large ;)

**Comment by Qix-** [comment]
`-c256` would have to be a standard GNU short-arg `-c 256`, of course.

---

So is the goal here to be able to disable this module completely from within an application?

**Comment by isaacs** [comment]
I decided that actually most of what this module does is not quite what I want, and what I actually want is a similar concept but almost entirely different behavior, so I just wrote what I wanted in JS rather than try to describe it  here :) https://github.com/isaacs/color-support

**Comment by Qix-** [comment]
@bolinfest I know it's been a while since you opened this ticket, but I'm curious as to what false positives you're referring to? If the user wants to use the flag `--colors=256`, then they should be able to force it (even if their immediate emulator doesn't support it).

I agree that with all of the environment variables being present, our hack-and-slash use of `argv` is a bit unnecessary; I'd be interested to see if we should re-evaluate its usage.

cc @sindresorhus

**Comment by plroebuck** [comment]
Would also prefer a split in the module's functionality. If possible, one method that _just_ queries the environment. Then another that wraps the former, adding the force logic (essentially what is now).

**Comment by isaacs** [comment]
@plroebuck that's essentially what https://github.com/isaacs/color-support is.

**Comment by sindresorhus** [comment]
I agree, we should expose a method that returns the color support without using the flags (as an addition, not instead of the current behavior).

**Comment by Qix-** [comment]
Since `chalk` is the primary consumer of this library, how would that look API-wise to have two versions for CLI flag checking?
